src/internal/m365/graph/middleware.go

```diff
@@ -252,93 +245,88 @@ func (mw RetryMiddleware) retryRequest(
-
-	// only retry under certain conditions:
-	// 1, there was an error.  2, the resp and/or status code match retriable conditions.
+	// only retry if all the following conditions are met:
+	// 1, there was a prior error.
+	// 2, the resp and/or status code match retriable conditions.
 	// 3, the request is retriable.
 	// 4, we haven't hit our max retries already.
-	if (priorErr != nil || mw.isRetriableRespCode(ctx, resp, statusCode)) &&
+	shouldRetry := (priorErr != nil || mw.isRetriableRespCode(ctx, resp)) &&
 		mw.isRetriableRequest(req) &&
-		executionCount < mw.MaxRetries {
-		executionCount++
+		executionCount < mw.MaxRetries
 
-		delay := mw.getRetryDelay(req, resp, exponentialBackoff)
-		cumulativeDelay += delay
+	if !shouldRetry {
+		return resp, stackReq(ctx, req, resp, priorErr).OrNil()
+	}
 
-		req.Header.Set(retryAttemptHeader, strconv.Itoa(executionCount))
+	executionCount++
 
-		timer := time.NewTimer(delay)
+	delay := mw.getRetryDelay(req, resp, exponentialBackoff)
+	cumulativeDelay += delay
 
-		select {
-		case <-ctx.Done():
-			// Don't retry if the context is marked as done, it will just error out
-			// when we attempt to send the retry anyway.
-			return resp, clues.Stack(ctx.Err()).WithClues(ctx)
+	req.Header.Set(retryAttemptHeader, strconv.Itoa(executionCount))
 
-		case <-timer.C:
-		}
+	timer := time.NewTimer(delay)
 
-		// we have to reset the original body reader for each retry, or else the graph
-		// compressor will produce a 0 length body following an error response such
-		// as a 500.
-		if req.Body != nil {
-			if s, ok := req.Body.(io.Seeker); ok {
-				_, err := s.Seek(0, io.SeekStart)
-				if err != nil {
-					return nil, Wrap(ctx, err, "resetting request body reader")
-				}
-			}
-		}
+	select {
+	case <-ctx.Done():
+		// Don't retry if the context is marked as done, it will just error out
+		// when we attempt to send the retry anyway.
+		return resp, clues.Stack(ctx.Err()).WithClues(ctx)
 
-		nextResp, err := pipeline.Next(req, middlewareIndex)
-		if err != nil && !IsErrTimeout(err) && !IsErrConnectionReset(err) {
-			return nextResp, stackReq(ctx, req, nextResp, err)
-		}
+	case <-timer.C:
+	}
 
-		return mw.retryRequest(
-			ctx,
-			pipeline,
-			middlewareIndex,
-			req,
-			nextResp,
-			executionCount,
-			cumulativeDelay,
-			exponentialBackoff,
-			err)
+	// we have to reset the original body reader for each retry, or else the graph
+	// compressor will produce a 0 length body following an error response such
+	// as a 500.
+	if req.Body != nil {
+		if s, ok := req.Body.(io.Seeker); ok {
+			if _, err := s.Seek(0, io.SeekStart); err != nil {
+				return resp, Wrap(ctx, err, "resetting request body reader")
+			}
+		}
 	}
 
-	if priorErr != nil {
-		return nil, stackReq(ctx, req, nil, priorErr)
+	nextResp, err := pipeline.Next(req, middlewareIndex)
+	if err != nil && !IsErrTimeout(err) && !IsErrConnectionReset(err) {
+		return nextResp, stackReq(ctx, req, nextResp, err)
 	}
```

Comment:
this also isn't directly modified in this PR but it does feel rather weird to check for a subset of errors and explicitly return on those rather than calling this function again with the response which could then handle it at the start of the function (or may already handle it)

Reply:
heh, no disagreement there.  I think that's a case which would fit more nicely into a func refactor, which we can handle some other time.